doc/modules/feature_selection.rst

```diff
@@ -47,6 +47,42 @@ so we can select using the threshold ``.8 * (1 - .8)``::
 As expected, ``VarianceThreshold`` has removed the first column,
 which has a probability :math:`p = 5/6 > .8` of containing a zero.
 
+.. _correlation_threshold:
+
+Remove features with high correlation
+=====================================
+
+:class:`CorrelationThreshold` is a feature selection technique that removes
+features until all pairwise correlations are below a certain threshold. As
+described in Section 3.5 of [KUHN]_, the algorithm is as follows:
+
+0. Remove features with zero variance because they would contribute to an
+   ill-defined correlation matrix.
+1. Calculate the correlation matrix of the features.
+2. Determine the two features associated with the largest absolute pairwise
+   correlation.
+3. Remove the feature with the higher average correlation with the rest of the
+   features.
+4. Repeat until there are no absolute correlations that are above the
+   threshold.
+
+The following example shows how one of the first highly correlated features
+are removed, while the uncorrelated feature remains:
+
+    >>> import numpy as np
+    >>> from sklearn.feature_selection import CorrelationThreshold
+    >>> X = np.array([[0.0, 1.0, 2.0], [1.1, 2.0, 3.0], [0.5, 10.1, 1.1]]).T
+    >>> selector = CorrelationThreshold(threshold=0.9)
+    >>> selector.fit_transform(X)
+    array([[ 1.1,  0.5],
+           [ 2. , 10.1],
+           [ 3. ,  1.1]])
+
+.. topic:: Reference
+
+    .. [KUHN] Max Kuhn and Kjell Johnson, "Applied Predictive Modeling",
+        Springer, 2013
+
```

Comment:
Can we have a reference to the original paper describing this method?

To me, it seems linked to stagewise regression, but I had never heard of it.

Reply:
This method is described in the `caret` paper (in R): https://www.jstatsoft.org/article/view/v028i05 (page 4-6), where they implemented this strategy in `findCorrelation`.

The earliest paper I can find: https://www.cs.waikato.ac.nz/~ml/publications/1997/Hall-LSmith97.pdf

Mark Hall wrote thesis on this topic here: https://www.cs.waikato.ac.nz/~mhall/thesis.pdf

A newer paper comparing it to other feature selection algorithms: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5738110/

Reply:
Thank you very much. I think that it would be useful to reference a paper, or at least an URL that can be downloaded by every one.